### PR TITLE
ISSUE #2026 - Comments UI improvement

### LINKED ISSUE #2026: Comments UI improvement
### Description
Improvement of comments UI in 3D Repo, allowing the user to filter comments, add references to a user or issue, quote comments, and have base support of markdown syntax.

Mockup: https://marvelapp.com/ad2hjh3/screen/68029558  

### Goals
- [x] A fresh layout for comments
- [x] Owner comments should be in a different colour to other comments.
- [x] Comment should be listed by date/time ascending, with scroll automatically scrolled to bottom by default
- [x] Add filter comments/system message/all to filter the comments
- [x] Avatar image of each user with outline colour matching its job colour
- [x] Show user card on hover
- [x] User reference should be in a emphasis colour and it should show user card on hover
- [x] Issue reference  should look like a hyperlink and show issue info on hover, go to issue if clicked (the back button on the issue card should then go back to issues list as normal, not back like a browser!)
- [x] Resource reference should look like a hyperlink, no need for tooltip and just act as if the user clicked on the resource on the resource list.
- [x] Add quote button to reference: when clicked, add reference link to comment (markdown - #res.<resourceID>)
- [x] Add quote button to comment: when clicked, add quote to comment (markdown)
- [x] Add reply button: when clicked, add reference to user (@username)
- [x] render markdown appropriately for comments
- [x] if user enters @xx... list of filtered users (users within teamspace) that matches the username should come up
- [x] if user enters #<num>... list of possible issues should come up
- [x] viewer quote/reply should be disabled if the user has no write access
- [x] comments should be sent as markdown and only be rendered visually on the fly
- [x] if the quoted text has an image, add domain name placeholder that is replaced on the fly depending on configuration
- [x] do not render references within quotes, it should be plain text + images only.

### PR DESCRIPTION
**Pull request by m1chalkubiak**
This fixes #2026

#### Description
- [x] A fresh layout for comments
- [x] Owner comments should be in a different colour to other comments.
- [x] Comment should be listed by date/time ascending, with scroll automatically scrolled to bottom by default
- [x] Add filter comments/system message/all to filter the comments
- [x] Avatar image of each user with outline colour matching its job colour
- [x] Show user card on hover
- [x] User reference should be in a emphasis colour and it should show user card on hover
- [x] Issue reference  should look like a hyperlink and show issue info on hover, go to issue if clicked (the back button on the issue card should then go back to issues list as normal, not back like a browser!)
- [x] Resource reference should look like a hyperlink, no need for tooltip and just act as if the user clicked on the resource on the resource list.
- [x] Add quote button to reference: when clicked, add reference link to comment (markdown - #res.<resourceID>)
- [x] Add quote button to comment: when clicked, add quote to comment (markdown)
- [x] Add reply button: when clicked, add reference to user (@username)
- [x] render markdown appropriately for comments
- [x] if user enters @xx... list of filtered users (users within teamspace) that matches the username should come up
- [x] if user enters #<num>... list of possible issues should come up
- [x] viewer quote/reply should be disabled if the user has no write access
- [x] comments should be sent as markdown and only be rendered visually on the fly
- [x] if the quoted text has an image, add domain name placeholder that is replaced on the fly depending on configuration
- [x] do not render references within quotes, it should be plain text + images only.

### CODE REVIEW

**Comment by carmenfan** [comment]
@m1chalkubiak can you look into fixing the travis error please?
You will have to merge in latest of staging if you haven't as #2029 is needed for a chrome driver issue

**Comment by m1chalkubiak** [comment]
@carmenfan  I've merged staging into this branch, fixed some missing items in package.json file. However, Travis is still crashing, currently on lack of some credentials, If i understand it correctly.

**Comment by carmenfan** [comment]
> @carmenfan I've merged staging into this branch, fixed some missing items in package.json file. However, Travis is still crashing, currently on lack of some credentials, If i understand it correctly.

That's fine thanks!
It's failing on the backend tests I added for the change on the endpoint. I couldn't test them earlier since travis was failing for other reason. I'll look into this

**Comment by carmenfan** [comment]
@m1chalkubiak found a few more and edited the previous comment. Looks good in general!

@sanmont3drepo  can you review the code?

**Comment by m1chalkubiak** [comment]
@carmenfan
Thank you very much for the detailed feedback, the list is long, but most can be done quickly. In the meantime, I also noticed that I broke comments on risks.
I didn't ask about it before, but I assume that new comments should work analogously on them.
I am already working on fixing it.

**Comment by m1chalkubiak** [comment]
> * Can we use a slightly darker colour here? It's a bit hard to read.. maybe 0.4 or 0.5 alpha?
>   image
> * [x]  Same with quotes:
>   image

> * [x] Darker colour on the lighter grey text as well
>   image

Hi @carmenfan 
I used 0.4  let me know if it should be darker.

All issues related to autosuggestion prompt displaying in weird places is connected with the same thing, which I currently trying to solve. I should have a fix for this tomorrow morning.

**Comment by carmenfan** [comment]
> > * Can we use a slightly darker colour here? It's a bit hard to read.. maybe 0.4 or 0.5 alpha?
> >   image
> > * [x]   Same with quotes:
> >   image
> 
> > * [x]  Darker colour on the lighter grey text as well
> >   image
> 
> Hi @carmenfan
> I used 0.4 let me know if it should be darker.

0.4 looks good! 👍 

> 
> All issues related to autosuggestion prompt displaying in weird places is connected with the same thing, which I currently trying to solve. I should have a fix for this tomorrow morning.

sounds good! thanks!

**Comment by carmenfan** [comment]
@m1chalkubiak  just spotted something, any chance of aligning them better? `Show:` seems to be aligned higher than comments which in turn is higher than the arrow:
image

**Comment by m1chalkubiak** [comment]
Hey @carmenfan I trying to solve issue with autosuggestion prompt displaying in weird places, by most of the day, but right now I came to a different idea how this can be done, but I need confirmation from You.

The previous solution assumed that they are displayed inside the container with a list of comments, but it is quite narrow and the library that I use seems to have some issue in calculating these positions in such a case.

So, Is this would be a problem if those promos would displays above comment container?

And to clarify what I mean I attaching screenshots:
Zrzut ekranu 2020-05-15 o 14 03 59
Zrzut ekranu 2020-05-15 o 14 07 41

**Comment by carmenfan** [comment]
> Hey @carmenfan I trying to solve issue with autosuggestion prompt displaying in weird places, by most of the day, but right now I came to a different idea how this can be done, but I need confirmation from You.
> 
> The previous solution assumed that they are displayed inside the container with a list of comments, but it is quite narrow and the library that I use seems to have some issue in calculating these positions in such a case.
> 
> So, Is this would be a problem if those promos would displays above comment container?
> 
> And to clarify what I mean I attaching screenshots:
> Zrzut ekranu 2020-05-15 o 14 03 59
> Zrzut ekranu 2020-05-15 o 14 07 41

@m1chalkubiak  yep, I think this is fine 👍

**Comment by m1chalkubiak** [comment]
> * Users who don't have avatar should get the default one (like teamspace list)

I show user initial in case of lack of avatar image as from what I have checked we have such logic in other places of the app, I hope this will be fine for You.

**Comment by m1chalkubiak** [comment]
> * [ ]  clicking on tagged issue from kanban is making the browser go to the model view instead of just loading that issue on the pop up

@carmenfan Can you guide me on what behavior would be the one that we expect?
Did linked issue should be rendered in the popup above the current one, or just reload data in the current one, unfortunately, we don't have any good way to go back to the previous one then.

**Comment by carmenfan** [comment]
> > * [ ]   clicking on tagged issue from kanban is making the browser go to the model view instead of just loading that issue on the pop up
> 
> @carmenfan Can you guide me on what behavior would be the one that we expect?
> Did linked issue should be rendered in the popup above the current one, or just reload data in the current one, unfortunately, we don't have any good way to go back to the previous one then.

@m1chalkubiak  it should just reload the issue data into the pop up. Don't worry about going backwards, like github we plan to add issue reference in the target issue.

**Comment by m1chalkubiak** [comment]
Hi @carmenfan 
Everything should work aligning with Your feedback.

> * [ ] Adding a comment should make the scroll go to the bottom

I fixed some condition and improve animation, now every time new item appears on the comments list container will be scrolled to the bottom.

> * [ ]  Expand arrow should be visible at all times. Seems like it is now on the top of the comments section instead.

This change required some tricks and took me a lot of time, currently, styling is changed depending on the details section expand state and I end up with using a setTimeout which is caused by the time of transition animation. I'm not proud of this solution, but It seems to work ok, and I want to be honest with You.
However, I would like to hear your opinion on this subject and if you decide that it does not work as expected I will try to solve it in a different way.

**Comment by m1chalkubiak** [comment]
@carmenfan 
I pushed quick fixes just to don't block You longer. But for both of them I would need to dig deeper what is to cause od them. 

> @m1chalkubiak this bug is stopping me from re-testing it as it's happening everytime I make a comment
> 
> * [ ]  I'm getting white-screened when I clicked on the issue I tested with last time:

It's about getting of user Avatar and can be related to any user listed on the comments list. 
It would be ideal if I could see exactly this model and see what users appear on the list there, it is possible that we have a case that I did not predict.

> * [ ]  I'm seeing an `Uncaught (in promise) TypeError: Cannot read property 'promise' of undefined` on the console on the kanban board, this might be a merge problem since I think @sanmont3drepo  cleaned up the promises on viewer when he finished off the refactoring:
>   image

I will look in commit history and try to find out how it should be solved.

**Comment by carmenfan** [comment]
> @carmenfan
> I pushed quick fixes just to don't block You longer. But for both of them I would need to dig deeper what is to cause od them.
> 
> > @m1chalkubiak this bug is stopping me from re-testing it as it's happening everytime I make a comment
> > 
> > * [ ]   I'm getting white-screened when I clicked on the issue I tested with last time:
> 
> It's about getting of user Avatar and can be related to any user listed on the comments list.
> It would be ideal if I could see exactly this model and see what users appear on the list there, it is possible that we have a case that I did not predict.
> 
> > * [ ]   I'm seeing an `Uncaught (in promise) TypeError: Cannot read property 'promise' of undefined` on the console on the kanban board, this might be a merge problem since I think @sanmont3drepo  cleaned up the promises on viewer when he finished off the refactoring:
> >   image
> 
> I will look in commit history and try to find out how it should be solved.

Actually this was my fault. I didn't restart the backend when I switch branch! It's working when I restarted.. sorry!

Though it did make me think... it's possible that the user is no longer in the teamspace if they were removed access.. so we may have to cater for a situation where we can't find the user's job or it's job colour

**Comment by m1chalkubiak** [comment]
@carmenfan 
> * [ ]  link to another issue is missing:
>   image

From the code perspective the only reason for no linked reference would be a lack of issue with such a number in this model.
So my first question would be: Is this issue (#1) currently exist at this model?
And if so, then there could be some bug in regex that I use to detect for references.

> * [ ]  The comments shouldn't populate from bottom, it should start to fill from top, just auto scroll to bottom if required:
> * [ ]  scrolling doesn't seem to work well when there's image:

So those two things are connected, at the were beginning of changing comments UI I realized that images inside them could be a challenge. The case is that images load asynchronously, depending of internet speed etc. At the moment when the list is drawn from the top, at the very beginning we would be scrolled down to the bottom of list, but after a moment when the pictures were loaded the container would expand but user scroll would stay in the same place. Changing the style to one that draws the list from the bottom makes it very easy to handle proper behavior in this situation.

However, the below point can solve this problem, as we could render placeholders for these images with the maximum allowed dimensions. Any subsequent "shrinkage" of the images should not be a problem.

Please let me know if an initial rendering of placeholders for images in comments is acceptable to you.

> * [ ]  Can we resize these images? set max height of issue screenshot to be 200px, and comment screenshots to be max height 100px? so it should look something like this:
>   image
>   instead of this:
>   image


> * [ ]  Collapsable panel is too long, the collpase button is being obscured:
>   image

I hope changing max height of issue screenshot should enough to fix it, or should I do something additional regarding this point?

**Comment by carmenfan** [comment]
@m1chalkubiak 

> From the code perspective the only reason for no linked reference would be a lack of issue with such a number in this model.
> So my first question would be: Is this issue (#1) currently exist at this model?
> And if so, then there could be some bug in regex that I use to detect for references.

I think it's a kanban thing, it works if i'm viewing the issues on the viewer

> Please let me know if an initial rendering of placeholders for images in comments is acceptable to you.

Yep. I think it's fine. If it's not too much trouble then I think it'd be good to have a place holder image/box incase the image takes longer than normal to load

> I hope changing max height of issue screenshot should enough to fix it, or should I do something additional regarding this point?
good point, let's leave it for now and see.

**Comment by m1chalkubiak** [comment]
> I think it's a kanban thing, it works if i'm viewing the issues on the viewer

I think I found the reason for this problem, it seems to me that now the references should link correctly.

File: frontend/modules/comments/comments.selectors.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+import { createSelector } from 'reselect';
+
+import { selectJobs } from '../jobs';
+
+export const selectUsersDomain = (state) => ({ ...state.comments });
+
+export const selectTeamspaceUsers = createSelector(
+	selectUsersDomain, selectJobs, (state, jobs) => state.users
+		.map(({ job, ...user }) => ({
```
[the remark this answers is not in the document]
Comment (m1chalkubiak) [reply frontend/modules/comments/comments.selectors.ts:26]: I'm using both, teamspaceUser list is needed for displaying the auto-suggested user's list during writing comment.

File: frontend/modules/comments/comments.sagas.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		const { data } = yield API.fetchUsers(teamspace);
+
+		yield put(CommentsActions.fetchUsersSuccess(data));
+	} catch (error) {
+		yield put(DialogActions.showEndpointErrorDialog('get', 'users', error));
+	}
+}
+
+export default function* CommentsSaga() {
+	yield takeEvery(CommentsTypes.FETCH_USERS, fetchUsers);
```
[the remark this answers is not in the document]
Comment (m1chalkubiak) [reply frontend/modules/comments/comments.sagas.ts:35]: I get it, it can be confusing, I will refactor it.

**Comment by m1chalkubiak** [comment]
@carmenfan @sanmont3drepo 
I added the last changes and thinks all of them should work fine.

Rendering placeholders took me a lot of time as I tried to use the previously used image library that should support image loading and displaying placeholders, but I couldn't make it work properly. And I ended up at writing handling for this on my own.

If you would find anything else that needs to be improved, please let me know

**Comment by m1chalkubiak** [comment]
> @m1chalkubiak sorry I know you're probably sick of this already but a few more things:

No problem @carmenfan , I will do it right away.

File: frontend/modules/comments/comments.selectors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/modules/comments/comments.selectors.ts:26]: I still don't understand, is the same set of data why are we dividing it, why arent we resuing the one with all the data and filtering those using a selector?

**Comment by carmenfan** [comment]
- [ ] New lines are not being formatted properly. notice how I typed this with new lines and they are being absorbed when it's being displayed:


- [ ] Text after double newline are escaping the quotes. If I quote the text i created for the previous bug I get this:
image
The last line is not being wrapped around the quote.

**Comment by m1chalkubiak** [comment]
> * [ ] The front end seems to be launchign an API request for every user reference within the comments.. Is this for the pop up card ? we could make it pull the data on demand instead of that's the case?:

@carmenfan It is strongly connected with Santiago's code suggestion, after checking it again and re-thinking over it, I have to agree with him. As we always have to have a full list of users to be able to handle their auto-suggestion, endpoint `/api/{teamspace}/members/{user}` seem to be completely unneeded on the FE side. At the beginning of this task, I focused too much on the existing logic for this module, while with the current structure a completely different approach can be used.

I'm currently in the middle of refactoring it should be ready tomorrow morning.

**Comment by m1chalkubiak** [comment]
@carmenfan
I end up refactoring, so now only one call is made for the whole list of teamspace members, and all details date are taken from there. 

I also handled the case when the user is no longer in the teamspace, then we will only have his username and only a simple placeholder avatar without any hover popup will be displayed.

File: frontend/modules/comments/comments.selectors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (m1chalkubiak) [reply frontend/modules/comments/comments.selectors.ts:37]: I refactored it so now all data are taken from one place. Thanks for pointing this out to me.

**Comment by m1chalkubiak** [comment]
Hi @carmenfan 
Sorry, I did not notice this comment yesterday because I am currently trying to concentrate completely on delivering "Issues / Risks UI update" feature to the end of next week.
Today I had some free time to fix this issue. However, because I am unable to add files locally to this project, I was not able to test it 100%. We have already tried to solve this issue once during work on SafetiBase v3, but we could not solve it, but if there is such a possibility I would contact Charence again at the beginning of next week to find a solution to this problem.

I have also added a small improvement adapting the message about the lack of items in the messages list depending on the selected filter type.

**Pull request merged**